connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/DataFrameNaFunctions.scala

```diff
@@ -0,0 +1,524 @@
+   *   df.na.replace("firstname" :: "lastname" :: Nil, Map("UNKNOWN" -> "unnamed"));
+   * }}}
+   *
+   * @param cols
+   *   list of columns to apply the value replacement. If `col` is "*", replacement is applied on
+   *   all string, numeric or boolean columns.
+   * @param replacement
+   *   value replacement map. Key and value of `replacement` map must have the same type, and can
+   *   only be doubles, strings or booleans. The map value can have nulls.
+   * @since 3.4.0
+   */
+  def replace[T](cols: Seq[String], replacement: Map[T, T]): DataFrame = {
+    buildReplaceDataFrame(Some(cols), buildReplacement(replacement))
+  }
+
+  private def buildReplaceDataFrame(
+      cols: Option[Seq[String]],
+      replacements: Iterable[NAReplace.Replacement]): DataFrame = {
+    sparkSession.newDataFrame { builder =>
+      val replaceBuilder = builder.getReplaceBuilder.setInput(root)
+      replaceBuilder.addAllReplacements(replacements.asJava)
+      cols.foreach(c => replaceBuilder.addAllCols(c.asJava))
+    }
+  }
+
+  private def buildReplacement[T](replacement: Map[T, T]): Iterable[NAReplace.Replacement] = {
+    // Convert the NumericType in replacement map to DoubleType,
+    // while leaving StringType, BooleanType and null untouched.
+    val replacementMap: Map[_, _] = replacement.map {
+      case (k, v: String) => (k, v)
+      case (k, v: Boolean) => (k, v)
+      case (k: String, null) => (k, null)
+      case (k: Boolean, null) => (k, null)
+      case (k, null) => (convertToDouble(k), null)
+      case (k, v) => (convertToDouble(k), convertToDouble(v))
+    }
+
+    // targetColumnType is either DoubleType, StringType or BooleanType,
+    // depending on the type of first key in replacement map.
+    // Only fields of targetColumnType will perform replacement.
+    val targetColumnType = replacement.head._1 match {
+      case _: jl.Double | _: jl.Float | _: jl.Integer | _: jl.Long => DoubleType
+      case _: jl.Boolean => BooleanType
+      case _: String => StringType
+    }
+
+    replacementMap.map { case (oldValue, newValue) =>
+      targetColumnType match {
+        case DoubleType =>
+          Replacement
+            .newBuilder()
+            .setOldValue(convertToLiteral(oldValue, DoubleType))
+            .setNewValue(convertToLiteral(newValue, DoubleType))
+            .build()
+        case BooleanType =>
+          Replacement
+            .newBuilder()
+            .setOldValue(convertToLiteral(oldValue, BooleanType))
+            .setNewValue(convertToLiteral(newValue, BooleanType))
+            .build()
+        case StringType =>
+          Replacement
+            .newBuilder()
+            .setOldValue(convertToLiteral(oldValue, StringType))
+            .setNewValue(convertToLiteral(newValue, StringType))
+            .build()
+      }
+    }
+  }
+
+  private def convertToLiteral(v: Any, dataType: DataType): GLiteral = dataType match {
```

Comment:
You don't need this. Just use `functions.lit(v).expr`. I looked at the SparkConnectPlanner and nulls handled properly in there.

Reply:
This should also significantly simplify the code in `replacementMap` and `fillMap`.

Reply:
OK, I have deleted convertToLiteral method